### PR TITLE
Created the load script for the pipeline | Improved extract and transform script

### PR DESCRIPTION
**Pull request by CharliedeanCS**
Worked on 3 files:

load.py
Created a load script that takes in the most recent collection of plant data and uploads it to the SQL Server
- Pylinted the code and improved efficieny

transform.py
Changed the transform file so it interacts with the data folder as this is where the data will be located

extract.py
Changed the extract script so it now extracts the plant id as inserts it into the database and ensures to keep the plant id from the API instead of creating its own index.

### CODE REVIEW

File: pipeline/load.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+"""Script to connect to an MSSQL database and insert plant data"""
```
Comment (zandersnow14) [inline pipeline/load.py:1]:
```suggestion
"""Script to connect to an MSSQL database and insert plant data."""
```

File: pipeline/load.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+from dotenv import load_dotenv
+from sqlalchemy import create_engine, sql, Connection
+import pandas as pd
+
+from transform import csv_to_data_frame
+
+
+def create_database_connection() -> Connection:
+    """Creates a database connection to the SQL Server"""
```
Comment (zandersnow14) [inline pipeline/load.py:14]:
```suggestion
    """Creates a database connection to the SQL Server."""
```

File: pipeline/load.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    engine = create_engine(
+        f"mssql+pymssql://{environ['DB_USERNAME']}:{environ['DB_PASSWORD']}@{environ['DB_HOST']}/?charset=utf8")
+
+    conn = engine.connect()
+
+    return conn
+
+
+def insert_into_recordings_table(connection: Connection, plant_data: pd.DataFrame) -> None:
+    """Inserts data from a pandas data frame into a SQL Server"""
```
Comment (zandersnow14) [inline pipeline/load.py:25]:
```suggestion
    """Inserts data from a pandas data frame into a SQL Server."""
```

File: pipeline/load.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            watered = plant[2]
+
+            connection.execute(sql.text("USE plants;"))
+
+            query = sql.text(
+                """INSERT INTO s_delta.recording (plant_id,soil_moisture,temperature,recording_taken,last_watered)
+                VALUES (:id,:soil,:temperature,:recording,:watered)""")
+            connection.execute(query, {"id": id, "soil": soil, "temperature": temperature, "recording": recording,
+                                       "watered": watered})
+        except:
```
Comment (zandersnow14) [inline pipeline/load.py]: Could be more specific with the exception here - what might go wrong here to require an exception?

File: pipeline/load.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            pass
```
Comment (shivani-63) [inline pipeline/load.py]: does this just leave values as null if they cannot be located using your try/except clause?

File: pipeline/transform.py [hunk 6]
```diff
@@ -66,26 +66,38 @@ def change_temp_and_moisture_to_two_dp(dataframe: pd.DataFrame) -> pd.DataFrame:
...
 
 
+def csv_to_data_frame(filename: str) -> pd.DataFrame:
+    """Converts a csv file to a pandas data frame."""
+
+    return pd.read_csv(filename)
+
+
+def upload_clean_csv_file(filename: str, plant_data: pd.DataFrame) -> None:
+    """Creates a new CSV file with the clean data"""
```
Comment (zandersnow14) [inline pipeline/transform.py:76]:
```suggestion
    """Creates a new CSV file with the clean data."""
```

File: pipeline/load.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+from os import environ
+
+from dotenv import load_dotenv
+from sqlalchemy import create_engine, sql, Connection
+import pandas as pd
+
+from transform import csv_to_data_frame
+
+
+def create_database_connection() -> Connection:
```
Comment (zandersnow14) [inline pipeline/load.py]: Quite nit picky, but maybe think about passing in a `config` argument here (so replace `environ` in this function with `config`, and then pass in environ in the `if __name__ == "__main__"` statement - makes things a bit neater in my head but just a thought.

File: pipeline/transform.py [hunk 8]
```diff
@@ -66,26 +66,38 @@ def change_temp_and_moisture_to_two_dp(dataframe: pd.DataFrame) -> pd.DataFrame:
...
-                               "Temperature", "Botanist Name", "Botanist Email", "Botanist Phone"])
+    plants = remove_rows_with_null(plants, ["Id", "Name", "Recording Taken", "Soil Moisture",
+                                            "Temperature", "Botanist Name", "Botanist Email", "Botanist Phone"])
 
     # Validate and round moisture and temperature values
-    df = check_soil_moisture_valid(df)
-    df = check_soil_temp_valid(df)
-    df = normalize_datetimes(df)
-    df = change_temp_and_moisture_to_two_dp(df)
+    plants = check_soil_moisture_valid(plants)
```
Comment (shivani-63) [inline pipeline/transform.py:97]: Really nice changes to variable names here, makes things clear and relevant!

File: pipeline/load.py (hunk 4 above)
Comment (CharliedeanCS) [reply pipeline/load.py]: I decided to remove the try-except clause as I realised it isn't needed and never runs into a problem at this part

File: pipeline/load.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+"""Script to connect to an MSSQL database and insert plant data"""
+
+
+from os import environ
```
Comment (zandersnow14) [inline pipeline/load.py]:
```suggestion
from os import environ, _Environ
```

File: pipeline/load.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+from dotenv import load_dotenv
+from sqlalchemy import create_engine, sql, Connection
+import pandas as pd
+
+from transform import csv_to_data_frame
+
+
+def create_database_connection(config) -> Connection:
```
Comment (zandersnow14) [inline pipeline/load.py]:
```suggestion
def create_database_connection(config: _Environ) -> Connection:
```

**Review by zandersnow14 — APPROVED** [review]

**Pull request merged**